common/interactivetx.c

```diff
@@ -0,0 +1,687 @@
+		}
+	}
+}
+
+static char *send_next(const tal_t *ctx,
+		       struct interactivetx_context *ictx,
+		       bool *finished)
+{
+	struct channel_id *cid = &ictx->channel_id;
+	struct psbt_changeset *set = ictx->change_set;
+	u64 serial_id;
+	u8 *msg = NULL;
+
+	*finished = false;
+
+	if(!set)
+		goto tx_complete;
+
+	if(tal_count(set->added_ins) != 0) {
+
+		const struct input_set *in = &set->added_ins[0];
+		struct bitcoin_outpoint outpoint;
+		u8 *prevtx;
+
+		if (!psbt_get_serial_id(&in->input.unknowns, &serial_id))
+			abort();
+
+		if(in->input.utxo)
+			prevtx = linearize_wtx(ctx,
+					       in->input.utxo);
+		else
+			return "PSBT needs the previous transaction set.";
+
+		memcpy(outpoint.txid.shad.sha.u.u8,
+		       in->tx_input.txhash,
+		       WALLY_TXHASH_LEN);
+
+		outpoint.n = in->tx_input.index;
+
+		msg = towire_tx_add_input(ctx, cid, serial_id,
+					  prevtx, in->tx_input.index,
+					  in->tx_input.sequence,
+					  NULL);
+
+		tal_arr_remove(&set->added_ins, 0);
+
+	}
+	else if (tal_count(set->rm_ins) != 0) {
+
+		if (!psbt_get_serial_id(&set->rm_ins[0].input.unknowns,
+					&serial_id))
+			abort();
+
+		msg = towire_tx_remove_input(ctx, cid, serial_id);
+
+		tal_arr_remove(&set->rm_ins, 0);
+	}
+	else if (tal_count(set->added_outs) != 0) {
+
+		struct amount_sat sats;
+		struct amount_asset asset_amt;
+		const struct output_set *out;
+		const u8 *script;
+
+		out = &set->added_outs[0];
+
+		if (!psbt_get_serial_id(&out->output.unknowns, &serial_id))
+			abort();
+
+		asset_amt = wally_tx_output_get_amount(&out->tx_output);
+		sats = amount_asset_to_sat(&asset_amt);
+		script = wally_tx_output_get_script(ctx, &out->tx_output);
+
+		msg = towire_tx_add_output(ctx,
+					   cid,
+					   serial_id,
+					   sats.satoshis, /* Raw: wire interface */
+					   script);
+
+		tal_arr_remove(&set->added_outs, 0);
+	}
+	else if (tal_count(set->rm_outs) != 0) {
+
+		if (!psbt_get_serial_id(&set->rm_outs[0].output.unknowns,
+					&serial_id))
+			abort();
+
+		msg = towire_tx_remove_output(ctx, cid, serial_id);
+
+		tal_arr_remove(&set->rm_outs, 0);
+	}
+	else { /* no changes to psbt */
+
+		goto tx_complete;
+	}
+
+	if(!msg)  {
+
+		return "Interactivetx::send_next failed to build a message";
+	}
+
+	peer_write(ictx->pps, msg);
+	return NULL;
+
+tx_complete:
+
+	*finished = true;
+
+	if(!ictx->pause_when_complete) {
+
+		msg = towire_tx_complete(ctx, cid);
+		peer_write(ictx->pps, msg);
+	}
+
+	return NULL;
+}
+
+char *process_interactivetx_updates(const tal_t *ctx,
+			 	    struct interactivetx_context *ictx,
+				    bool *received_tx_complete)
+{
+	assert(NUM_TX_MSGS == INTERACTIVETX_NUM_TX_MSGS);
+
+	if(ictx->current_psbt == NULL)
+		ictx->current_psbt = create_psbt(ctx, 0, 0, 0);
+
+	bool we_complete = false, they_complete = false;
+	u8 *msg;
+	char *error = NULL;
+
+	if(received_tx_complete)
+		they_complete = *received_tx_complete;
+
+	/* Build change_set and handle PSBT variables */
+
+	ictx->change_set = NULL;
```

Comment:
It's meant to be only used internally for a single run of process_interactivetx_updates. The comment in the struct definition hopefully makes that clear
```
	/* Internal cached change set */
	struct psbt_changeset *change_set;
```

Any value already in it would be from a prior run and should be ignored.

Reply:
```suggestion
	ictx->change_set = tal_free(ictx->change_set);
```